### PR TITLE
Support data ingestion for generating offline segment in one pass

### PR DESCRIPTION
**Pull request by jackjlli**
## Description
Currently when generating an offline segment, raw data has to be traversed two times; one time for gathering stats in `RecordReaderSegmentCreationDataSource`, another time for ingesting actual data.

In some scenarios where the input data is streaming data (like Dataset<Row> in spark), since the raw data has to be traversed two times, the raw data has to be kept in the memory and thus squeeze the memory used for generating pinot segment. 

In this PR, we introduce a new way of ingesting offline data by traversing the raw data only once. Similarly to mutable realtime segment, a simplified class called `IntermediateSegment` will be initialized to gather all the data into dictionary and forwarded index. After all the records are ingested, `IntermediateSegmentRecordReader` will be passed into `SegmentIndexCreationDriver` with `SegmentGeneratorConfig`, and the final offline segment will be built like this:
```
    // Build the segment from intermediate segment.
    SegmentIndexCreationDriverImpl driver = new SegmentIndexCreationDriverImpl();
    driver.init(segmentGeneratorConfig, intermediateSegmentRecordReader);
    driver.build();
```

This mechanism helps greatly reduce the memory pressure in some environment. E.g. if pinot segments need to be generated directly from spark executors when data is loaded as iterator, the raw data doesn't have to load into memory two times.

## Upgrade Notes
Does this PR prevent a zero down-time upgrade? (Assume upgrade order: Controller, Broker, Server, Minion)
* [ ] Yes (Please label as **<code>backward-incompat</code>**, and complete the section below on Release Notes)

Does this PR fix a zero-downtime upgrade introduced earlier?
* [ ] Yes (Please label this as **<code>backward-incompat</code>**, and complete the section below on Release Notes)

Does this PR otherwise need attention when creating release notes? Things to consider:
- New configuration options
- Deprecation of configurations
- Signature changes to public methods/interfaces
- New plugins added or old plugins removed
* [ ] Yes (Please label this PR as **<code>release-notes</code>** and complete the section on Release Notes)
## Release Notes
If you have tagged this as either backward-incompat or release-notes,
you MUST add text here that you would like to see appear in release notes of the
next release.

If you have a series of commits adding or enabling a feature, then
add this section only in final commit that marks the feature completed.
Refer to earlier release notes to see examples of text

## Documentation
If you have introduced a new feature or configuration, please add it to the documentation as well.
See https://docs.pinot.apache.org/developers/developers-and-contributors/update-document

### CODE REVIEW

**Review by mayankshriv — COMMENTED** [review]
I think this is still two pass, right? The second pass is on IntermediateSegment.
Could you elaborate a bit on where's the memory pressure coming from in the existing implementation? And how does this approach solve it (IntermediateSegment) does need its own storage.

**Comment by jackjlli** [comment]
> I think this is still two pass, right? The second pass is on IntermediateSegment.
> Could you elaborate a bit on where's the memory pressure coming from in the existing implementation? And how does this approach solve it (IntermediateSegment) does need its own storage.

This is still two passes, but only one pass on the raw data.
E.g. in spark job where the data type is on row basis (like Dataset\<Row\>, RDD\<Row\>, Dataset\<T\>, RDD\<T\>), in order to  traverse the record, the API forEachPartition(iterator) is required to be called. While iterator can only provide the data once, thus the raw data has to be cached in the same executor in order to traverse it two times, if we stick on to the existing segment generation code.

The below lists out the possible solutions:
1) cache all the raw data into executor memory in order to reuse the same data (using existing code of segment generation)
2) traverse the raw data once and gathers the intermediate stats, and create offline segment based on the intermediate results (current PR)
3) call forEachPartition(iterator) two times; gathering stats to an extra DF/RDD in the 1st iteration, and ingesting actual raw data based on the former DF/RDD in the 2nd round (code refactor still needed, and one extra DF/RDD involved)

The Approach 1) needs to load all the raw data into the singe executor before initializing the `SegmentIndexCreationDriver`( raw data traversal is encapsulated in that class). The Approach 2) and 3) basically have the same idea that spilt the current code into two parts, while the difference is that 2) works on a stats collector and 3) works on the raw data (one extra data traversal).

**Comment by mayankshriv** [comment]
Thanks for the details @jackjlli, could you also is IntermediateSegment better than existing MutableSegment? For example, you could stream input data to MutableSegment and flush it as needed. This also solves multiple problems:
- Common code base for offline and RT segment generation (at least for the streaming part).
- Sorting can now be done for offline within SegmentGeneration, instead of having users to explicitly do so.
- Auto segment sizing that happens in RT will can also be done with offline now.

Thoughts @jackjlli @Jackie-Jiang?

**Comment by mcvsubbu** [comment]
> Thanks for the details @jackjlli, could you also is IntermediateSegment better than existing MutableSegment? For example, you could stream input data to MutableSegment and flush it as needed. This also solves multiple problems:
> 
> * Common code base for offline and RT segment generation (at least for the streaming part).
> * Sorting can now be done for offline within SegmentGeneration, instead of having users to explicitly do so.
> * Auto segment sizing that happens in RT will can also be done with offline now.
> 
> Thoughts @jackjlli @Jackie-Jiang?

I think this is a good idea to explore, but I suspect memory utilization on the offline side may go up significantly.

Also, the auto-segment sizing in realtime is implemented (in the controller) by learning the history of segments already completed. For offline generation, if we can keep a history or some learning mechanism, then it may be possible to implement approximate segment sizing algorithms -- whether we use MutableSegment to build segments or not.

**Comment by jackjlli** [comment]
> > Thanks for the details @jackjlli, could you also is IntermediateSegment better than existing MutableSegment? For example, you could stream input data to MutableSegment and flush it as needed. This also solves multiple problems:
> > 
> > * Common code base for offline and RT segment generation (at least for the streaming part).
> > * Sorting can now be done for offline within SegmentGeneration, instead of having users to explicitly do so.
> > * Auto segment sizing that happens in RT will can also be done with offline now.
> > 
> > Thoughts @jackjlli @Jackie-Jiang?
> 
> I think this is a good idea to explore, but I suspect memory utilization on the offline side may go up significantly.
> 
> Also, the auto-segment sizing in realtime is implemented (in the controller) by learning the history of segments already completed. For offline generation, if we can keep a history or some learning mechanism, then it may be possible to implement approximate segment sizing algorithms -- whether we use MutableSegment to build segments or not.

1. Yes, memory utilization will go up significantly, that's why I didn't directly use `MutableSegment` but `IntermediateSegment` as the intermediate container here.  In fact, both `IntermediateSegment` and `MutableSegment` share the common minimal piece of logic, which is that both have forwarded index. The slight difference is that `MutableSegment` will have all the indices (if applicable) like inverted index, text index, etc, for querying purposes. `IntermediateSegment` just keep the minimal component like dictionary.
2. Plus, if we want partitioning/ sorting, these steps can be done in the platform (like mapreduce, spark) before converting the raw data. In fact, we've already had that logic in LinkedIn. Once this PR is committed, we can consider open sourcing that spark code as well.
3. Auto-segment size is a good idea that historical data can be used to predict the cardinality or buffer size. While offline segment generation is not always done on the same machine, the historical data would be meaningless if they cannot be reused. If historical data is from controller, then all the worker machines have to query pinot controller simultaneously in order to get the historical data, which could bring huge amount of queries to controller. That's why I didn't bring it here in this PR. We can always add it to `IntermediateSegment` in the future, since the structure between `IntermediateSegment` and `MutableSegment` are pretty much the same.

All 3 points above are really good features, but it'd be too much to be in a single PR. It'd be good if we can leave room for those features and pick them up in the following PRs if applicable.

File: pinot-core/src/main/java/org/apache/pinot/core/indexsegment/mutable/IntermediateSegment.java [hunk 1]
```diff
@@ -0,0 +1,375 @@
...
+      _partitionFunction = PartitionFunctionFactory
+          .getPartitionFunction(segmentPartitionConfig.getFunctionName(_partitionColumn),
+              segmentPartitionConfig.getNumPartitions(_partitionColumn));
+    } else {
+      _partitionColumn = null;
+      _partitionFunction = null;
+    }
+
+    IndexLoadingConfig indexLoadingConfig = new IndexLoadingConfig(null, _tableConfig);
+    boolean offHeap = indexLoadingConfig.isRealtimeOffHeapAllocation();
```
Comment (snleee) [inline pinot-core/src/main/java/org/apache/pinot/core/indexsegment/mutable/IntermediateSegment.java:111]: `IntermediateSegment` should not use the config from real-time table config. You can just pick to use MMAP.

File: pinot-core/src/main/java/org/apache/pinot/core/indexsegment/mutable/IntermediateSegment.java [hunk 2]
```diff
@@ -0,0 +1,375 @@
...
+            buildAllocationContext(_segmentName, column, V1Constants.Indexes.UNSORTED_SV_FORWARD_INDEX_FILE_EXTENSION);
+        forwardIndex = new FixedByteSVMutableForwardIndex(true, FieldSpec.DataType.INT, _capacity, _memoryManager,
+            allocationContext);
+      } else {
+        // Multi-value dictionary-encoded forward index
+        String allocationContext =
+            buildAllocationContext(_segmentName, column, V1Constants.Indexes.UNSORTED_MV_FORWARD_INDEX_FILE_EXTENSION);
+        // TODO: Start with a smaller capacity on FixedByteMVForwardIndexReaderWriter and let it expand
+        forwardIndex = new FixedByteMVMutableForwardIndex(MAX_MULTI_VALUES_PER_ROW,
+            indexLoadingConfig.getRealtimeAvgMultiValueCount(), _capacity, Integer.BYTES, _memoryManager,
```
Comment (snleee) [inline pinot-core/src/main/java/org/apache/pinot/core/indexsegment/mutable/IntermediateSegment.java:163]: Same here. We should not depend on the realtime table config

File: pinot-core/src/main/java/org/apache/pinot/core/indexsegment/mutable/IntermediateSegment.java [hunk 3]
```diff
@@ -0,0 +1,375 @@
...
+  private final int _capacity = DEFAULT_CAPACITY;
+  private volatile int _numDocsIndexed = 0;
+
+  public IntermediateSegment(SegmentGeneratorConfig segmentGeneratorConfig) {
+    _segmentGeneratorConfig = segmentGeneratorConfig;
+    _schema = segmentGeneratorConfig.getSchema();
+    _tableConfig = segmentGeneratorConfig.getTableConfig();
+    _segmentName = _segmentGeneratorConfig.getSegmentName();
+
+    Collection<FieldSpec> allFieldSpecs = _schema.getAllFieldSpecs();
```
Comment (snleee) [inline pinot-core/src/main/java/org/apache/pinot/core/indexsegment/mutable/IntermediateSegment.java:92]: What is the purpose of L92-95? It does the multiple conversions `Collection<FieldSpec> -> List<FieldSpec> -> Collection<FieldSpec>`. Why we need to make this unmodifiable?

File: pinot-core/src/main/java/org/apache/pinot/core/indexsegment/mutable/IntermediateSegment.java [hunk 4]
```diff
@@ -0,0 +1,375 @@
...
+    List<FieldSpec> physicalFieldSpecs = new ArrayList<>(allFieldSpecs.size());
+    physicalFieldSpecs.addAll(allFieldSpecs);
+    Collection<FieldSpec> physicalFieldSpecs1 = Collections.unmodifiableCollection(physicalFieldSpecs);
```
Comment (snleee) [inline pinot-core/src/main/java/org/apache/pinot/core/indexsegment/mutable/IntermediateSegment.java:95]: Let's rename this variable `physicalFieldSpecs` -> `unmodifiablePhysicalFieldSpecs`

**Review by snleee — COMMENTED** [review]
1st round of review.

Can you post some profiling data, if you have one? I would like to see the memory consumption comparison of building a segment with `IntermediateSegmentRecordReader` and `AvroRecordReader`.

File: pinot-core/src/main/java/org/apache/pinot/core/indexsegment/mutable/IntermediateSegment.java (hunk 1 above)
Comment (jackjlli) [reply pinot-core/src/main/java/org/apache/pinot/core/indexsegment/mutable/IntermediateSegment.java]: Updated to straightly use offheap

File: pinot-core/src/main/java/org/apache/pinot/core/indexsegment/mutable/IntermediateSegment.java (hunk 2 above)
Comment (jackjlli) [reply pinot-core/src/main/java/org/apache/pinot/core/indexsegment/mutable/IntermediateSegment.java]: Removed.

File: pinot-core/src/main/java/org/apache/pinot/core/indexsegment/mutable/IntermediateSegment.java (hunk 3 above)
Comment (jackjlli) [reply pinot-core/src/main/java/org/apache/pinot/core/indexsegment/mutable/IntermediateSegment.java:95]: Updated.

File: pinot-core/src/main/java/org/apache/pinot/core/indexsegment/mutable/IntermediateSegment.java (hunk 4 above)
Comment (jackjlli) [reply pinot-core/src/main/java/org/apache/pinot/core/indexsegment/mutable/IntermediateSegment.java]: We don't need two copies here. Updated.

File: pinot-core/src/main/java/org/apache/pinot/core/segment/creator/IntermediateSegmentSegmentCreationDataSource.java [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+
+
+  public IntermediateSegmentSegmentCreationDataSource(IntermediateSegmentRecordReader intermediateSegmentRecordReader) {
+    _intermediateSegmentRecordReader = intermediateSegmentRecordReader;
+    _intermediateSegment = _intermediateSegmentRecordReader.getIntermediateSegment();
+  }
+
+  @Override
+  public SegmentPreIndexStatsContainer gatherStats(StatsCollectorConfig statsCollectorConfig) {
+
```
Comment (snleee) [inline pinot-core/src/main/java/org/apache/pinot/core/segment/creator/IntermediateSegmentSegmentCreationDataSource.java:42]: remove line

File: pinot-core/src/main/java/org/apache/pinot/core/realtime/converter/stats/MutableNoDictionaryColStatistics.java [hunk 6]
```diff
@@ -28,11 +28,11 @@
 import static org.apache.pinot.core.common.Constants.UNKNOWN_CARDINALITY;
 
 
-public class RealtimeNoDictionaryColStatistics implements ColumnStatistics {
+public class MutableNoDictionaryColStatistics implements ColumnStatistics {
```
Comment (snleee) [inline pinot-core/src/main/java/org/apache/pinot/core/realtime/converter/stats/MutableNoDictionaryColStatistics.java:31]: Do we ever read this from the controller? Can you confirm that this is only used by the server?

File: pinot-core/src/main/java/org/apache/pinot/core/realtime/converter/stats/MutableColumnStatistics.java [hunk 7]
```diff
@@ -33,15 +33,15 @@
  *
  * TODO: Gather more info on the fly to avoid scanning the segment
  */
-public class RealtimeColumnStatistics implements ColumnStatistics {
+public class MutableColumnStatistics implements ColumnStatistics {
```
Comment (snleee) [inline pinot-core/src/main/java/org/apache/pinot/core/realtime/converter/stats/MutableColumnStatistics.java:36]: Do we ever read this from the controller? Can you confirm that this is only used by the server?

If this is used by multiple components, renaming may cause incompatibility during the deployement.

File: pinot-core/src/main/java/org/apache/pinot/core/data/readers/IntermediateSegmentRecordReader.java [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+import java.io.IOException;
+import java.util.Set;
+import javax.annotation.Nullable;
+import org.apache.pinot.core.indexsegment.mutable.IntermediateSegment;
+import org.apache.pinot.spi.data.readers.GenericRow;
+import org.apache.pinot.spi.data.readers.RecordReader;
+import org.apache.pinot.spi.data.readers.RecordReaderConfig;
+
+
+public class IntermediateSegmentRecordReader implements RecordReader {
```
[the remark this answers is not in the document]
Comment (jackjlli) [reply pinot-core/src/main/java/org/apache/pinot/core/data/readers/IntermediateSegmentRecordReader.java:31]: Sorting can be done from either the reader or the input data. We can do that in later PR.

File: pinot-core/src/main/java/org/apache/pinot/core/realtime/converter/stats/MutableColumnStatistics.java (hunk 7 above)
Comment (jackjlli) [reply pinot-core/src/main/java/org/apache/pinot/core/realtime/converter/stats/MutableColumnStatistics.java:36]: The implementation of `ColumnStatistics` is used during segment creation. As long as there is no protocol change, I don't think renaming will cause incompatibility during the deployment.

File: pinot-core/src/main/java/org/apache/pinot/core/indexsegment/generator/SegmentGeneratorConfig.java [hunk 9]
```diff
@@ -679,4 +680,12 @@ public boolean isFailOnEmptySegment() {
   public void setFailOnEmptySegment(boolean failOnEmptySegment) {
     _failOnEmptySegment = failOnEmptySegment;
   }
+
+  public boolean isIntermediateSegmentRecordReader() {
```
[the remark this answers is not in the document]
Comment (jackjlli) [reply pinot-core/src/main/java/org/apache/pinot/core/indexsegment/generator/SegmentGeneratorConfig.java]: Good point. Removed this config.

File: pinot-core/src/main/java/org/apache/pinot/core/data/readers/IntermediateSegmentRecordReader.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (snleee) [reply pinot-core/src/main/java/org/apache/pinot/core/data/readers/IntermediateSegmentRecordReader.java:31]: Sure, let's address it when we unify all the offline ingestion path to use intermediate segment

**Review by snleee — APPROVED** [review]
LGTM

**Pull request merged**